### PR TITLE
SOLR-16477: Collection RENAME api creates broken alias

### PR DESCRIPTION
**Pull request by stillalex**
https://issues.apache.org/jira/browse/SOLR-16477

# Description

Collection RENAME api creates broken alias, named as source pointing to collection named as target.

# Solution

Revert params to create alias as 'target' pointing to source collection.

# Tests

Unit tests for Aliases class covering the rename.
I had a look at a failing test under CollectionsAPISolrJTest and had to tweak the assertions a bit, would really appreciate some input if the changes make sense or not.
Just to add to this, I'm currently not sure if a client going through SolrJ can issue a rename which has a different path than the api I changed, where the expectation is different, or this test is just emulating the code as it was written and effectively mirroring the bug. [Edit] managed to reproduce this bug via SolrJ as well, see Jira ticket for code sample.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

File: solr/CHANGES.txt [hunk 1]
```diff
@@ -65,6 +65,8 @@ Optimizations
 Bug Fixes
 ---------------------
 
+* SOLR-16477: Collection RENAME api creates broken alias
```
Comment (risdenk) [inline solr/CHANGES.txt:68]: nit: usually this goes at the bottom of the list.

File: solr/solrj/src/test/org/apache/solr/common/cloud/AliasesTest.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+import org.apache.solr.SolrTestCase;
+import org.apache.solr.common.SolrException;
+
+/** Unit test for {@link Aliases} */
+public class AliasesTest extends SolrTestCase {
+
+  public void testCloneWithRenameNullBefore() {
+    Aliases aliases = Aliases.EMPTY;
+    try {
+      aliases.cloneWithRename(null, "alias1");
```
Comment (risdenk) [inline solr/solrj/src/test/org/apache/solr/common/cloud/AliasesTest.java:30]: `assertThrows` would be better for this test?

File: solr/solrj/src/test/org/apache/solr/common/cloud/AliasesTest.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+      assertEquals(SolrException.ErrorCode.BAD_REQUEST.code, e.code());
+    }
+  }
+
+  public void testCloneWithRenameEmptyBefore() {
+    Aliases aliases = Aliases.EMPTY;
+    try {
+      aliases.cloneWithRename("", "alias1");
+      fail("null value should not be allowed");
+    } catch (SolrException e) {
```
Comment (risdenk) [inline solr/solrj/src/test/org/apache/solr/common/cloud/AliasesTest.java:42]: `assertThrows` would be better for this test?

File: solr/solrj/src/test/org/apache/solr/common/cloud/AliasesTest.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+      assertEquals(SolrException.ErrorCode.BAD_REQUEST.code, e.code());
+    }
+  }
+
+  public void testCloneWithRenameEmptyAfter() {
+    Aliases aliases = Aliases.EMPTY;
+    try {
+      aliases.cloneWithRename("alias0", "");
+      fail("null value should not be allowed");
+    } catch (SolrException e) {
```
Comment (risdenk) [inline solr/solrj/src/test/org/apache/solr/common/cloud/AliasesTest.java:52]: `assertThrows` would be better for this test?

**Review by risdenk — CHANGES_REQUESTED** [review]
I left some minor feedback. I did not look over this in any detail or if this change makes sense. Just saw some style stuff.

File: solr/CHANGES.txt (hunk 1 above)
Comment (stillalex) [reply solr/CHANGES.txt]: moved it to the bottom

**Comment by stillalex** [comment]
thanks for taking a look @risdenk! updated PR based on your observations so far.

**Comment by stillalex** [comment]
There is some code in the Aliases class I don't get:

```
public static Map<String, List<String>> convertMapOfCommaDelimitedToMapOfList(
      Map<String, String> collectionAliasMap) {
    return collectionAliasMap.entrySet().stream()
        .collect(
            Collectors.toMap(
                Map.Entry::getKey,
                e -> splitCollections(e.getValue()),
                (a, b) -> {
                  throw new IllegalStateException(
                      String.format(Locale.ROOT, "Duplicate key %s", b));
                },
                LinkedHashMap::new));
  }
```

if the starting point for the stream is a `Map<>`, the 'merge keys' part where it checks for collisions`(a, b) -> IllegalStateException` would never be possible, as the keys are already guaranteed to be unique by the map contract. am I reading this wrong?

**Comment by gerlowskija** [comment]
> if the starting point for the stream is a Map<>, the 'merge keys' part where it checks for collisions(a, b) -> IllegalStateException would never be possible

In this case, I think that's true - we're not really "mapping" the keys in the sense this method allows, we're just taking them unmodified from an existing Map.  I can imagine cases where the merge/collection function would be necessary, but it looks redundant here to me 🤷

**Review by risdenk — APPROVED** [review]

**Comment by stillalex** [comment]
thank you @gerlowskija for venturing a guess :) even though benign, having this dead code around adds a bit of cognitive complexity when trying to understand what it actually does.

**Review by risdenk — COMMENTED** [review]
@stillalex I noticed in `Aliases.java` line 322 - still has `throw new NullPointerException("Alias name cannot be null");`

Is it possible this should be addressed as well? I think based on your new tests it might be possible to add a test to fix this case as well? null for an alias?

**Comment by risdenk** [comment]
@stillalex maybe I'm old school, but I think the method looks better written without streams:

```
public static Map<String, List<String>> convertMapOfCommaDelimitedToMapOfList(
      Map<String, String> collectionAliasMap) {
    Map<String, List<String>> map = new LinkedHashMap<>();
    for (Map.Entry<String, String> entry : collectionAliasMap.entrySet()) {
      map.put(entry.getKey(), splitCollections(entry.getValue()));
    }
    return map;
  }
```

assuming I converted that back correctly.

**Comment by stillalex** [comment]
> @stillalex I noticed in `Aliases.java` line 322 - still has `throw new NullPointerException("Alias name cannot be null");`
> 
> Is it possible this should be addressed as well? I think based on your new tests it might be possible to add a test to fix this case as well? null for an alias?

I avoided doing too much cleanup/refactoring outside of the reported bug, but I can include this change as well in the PR.

**Comment by risdenk** [comment]
> I avoided doing too much cleanup/refactoring outside of the reported bug, but I can include this change as well in the PR.

Agreed makes sense, but I think in this case if you add a test case for `SolrException e = assertThrows(SolrException.class, () -> aliases.cloneWithRename("alias1", null));` you might trigger that other NPE case?  At least from a quick glance that seemed related to the other NPE test case you added.

**Comment by stillalex** [comment]
@risdenk added the NPE change + some more tests. shouldn't be any breaking test, but I am running the complete test suite again locally for validation.

**Review by risdenk — APPROVED** [review]

**Comment by risdenk** [comment]
Thanks @stillalex!
